Experimental: Navigation block with the Interactivity API

SantosGuillamot opened the pull request:
## What?

This pull request adds an experimental option to toggle between the Navigation block using [the Interactivity API](https://make.wordpress.org/core/2023/03/30/proposal-the-interactivity-api-a-better-developer-experience-in-building-interactive-blocks/) and the old scripts that use [`micromodal`](https://micromodal.vercel.app/):

* Option deactivated: The navigation block keeps working exactly as it currently does.
* Option activated: The JS files are substituted for the Interactivity API logic, and it adds the appropriate directives to the markup using the HTML Tag Processor.

I recorded this video explaining in detail the purpose of the pull request and how it has been implemented:

https://www.loom.com/share/c07afacb4e464de8a8e7b253930dcd1b

## Why?

This pull request is part of the [proposal to use the Interactivity API](https://make.wordpress.org/core/2023/03/30/proposal-the-interactivity-api-a-better-developer-experience-in-building-interactive-blocks/) as a standard for the blocks frontend. The main goal is to start testing real blocks with this API, which could also serve as an example, and see what they would look like.

In terms of performance, it is true that, if you only use the Interactivity API for the navigation block, it sends a few more kBs to the client than the current approach (12kBs vs 4kBs). However, if more blocks use the Interactivity API, they would reuse those 12kBs and would result in fewer kBs overall. There are already ongoing pull requests that aim to use this code like the [Lightbox in images](https://github.com/WordPress/gutenberg/pull/49621). And there could be more core blocks in the future using it.

Additionally, this would prevent the block from depending on external libraries like `micromodal`.

## How?

* It creates an experimental setting to toggle between the Navigation block using the Interactivity API and the old scripts (micromodal).
* Depending on that setting, it enqueues the relevant scripts and adds the Interactivity API directives to the markup using the Tag Processor. If we finally decide to use the Interactivity API for the navigation block, we can move the directives directly to the markup.
* It adds the logic needed to replicate what `micromodal` is doing.

There are several options in the navigation block. These are the use cases taken into account to replicate the current behavior:

1. **Overlay menu**: The menu modal that opens after clicking the hamburger button.
![Screenshot 2023-05-09 at 10 03 24](https://user-images.githubusercontent.com/34552881/237033562-4430e89d-f4b5-4592-95c1-72defcea8823.png)

* Overlay menu opens on click on "open menu" button.
* Overlay menu closes on click on "close menu" button.
* Overlay menu closes when ESC key is pressed.
* Overlay menu focuses on first element after opening.
* Overlay menu traps focus.

2. **Submenus (open on click)**: Submenus when "Open on click" option is activated:
![Screenshot 2023-05-09 at 10 06 32](https://user-images.githubusercontent.com/34552881/237034241-914d35b2-117f-4134-89f4-2401c18c6cca.png)

* Submenu opens on click.
* Nested submenu opens on click.
* Submenu closes on click outside.
* Nested submenu closes on click outside.
* Submenu closes on ESC key.
* Nested submenu closes on ESC key
* Submenu closes on tab outside submenu.
* Only nested submenu closes on tab outside.

3. **Submenus (Arrow setting)**: Submenus when "Open on click" is deactivated and "Show arrow" is activated.
![Screenshot 2023-05-09 at 10 11 22](https://user-images.githubusercontent.com/34552881/237035480-f42ed39e-c663-4d8b-b71c-0ed77587b1ad.png)

* Submenu opens on click in the arrow. The text is currently a link.
* The rest of the aspects should behave the same way as with the other option.

4. **Page list block**: The Page list block, which should behave as the submenus.

I tried to cover all the use cases, but let me know if you are missing anything.

## Testing

Basically, we have to ensure that the Navigation block, the navigation submenu, and the page list behave exactly the same when the option is activated/deactivated.

* Toggle the Experimental option in Gutenberg -> Experiments:
  * Check that the mentioned experiences (and anything not mentioned) work as expected.
  * Check that it is only sending the Interactivity API scripts.
  * Change the settings of the navigation block to ensure it works fine in all cases.
  * If the Overlay menu is set to off and the "Show arrow" is set to false, the scripts shouldn't be loaded.

As part of this pull request, I added [e2e tests for the navigation](https://github.com/WordPress/gutenberg/pull/50462) to check all those user experiences mentioned previously. With both options tests are passing, but maybe there are more use cases not contemplated there.

luisherranz committed:
Add interactivity runtime

Co-authored-by: David Arenas <david.arenas@automattic.com>

luisherranz committed:
Add interactivity runtime

Co-authored-by: David Arenas <david.arenas@automattic.com>

luisherranz committed:
Add interactivity runtime

Co-authored-by: David Arenas <david.arenas@automattic.com>

luisherranz committed:
Add it to the image block

This is still pretty basic, just to check that it works.

Co-authored-by: David Arenas <david.arenas@automattic.com>

luisherranz committed:
Add it to the image block

This is still pretty basic, just to check that it works.

Co-authored-by: David Arenas <david.arenas@automattic.com>

luisherranz committed:
Add it to the image block

This is still pretty basic, just to check that it works.

Co-authored-by: David Arenas <david.arenas@automattic.com>

luisherranz committed:
Add a separate webpack config

Co-authored-by: David Arenas <david.arenas@automattic.com>

luisherranz committed:
Add a separate webpack config

Co-authored-by: David Arenas <david.arenas@automattic.com>

luisherranz committed:
Add a separate webpack config

Co-authored-by: David Arenas <david.arenas@automattic.com>

luisherranz committed:
Make sure the runtime is imported only once

Co-authored-by: David Arenas <david.arenas@automattic.com>

luisherranz committed:
Make sure the runtime is imported only once

Co-authored-by: David Arenas <david.arenas@automattic.com>

luisherranz committed:
Make sure the runtime is imported only once

Co-authored-by: David Arenas <david.arenas@automattic.com>

luisherranz committed:
Use sideEffects instead of init

Co-authored-by: David Arenas <david.arenas@automattic.com>

luisherranz committed:
Use sideEffects instead of init

Co-authored-by: David Arenas <david.arenas@automattic.com>

luisherranz committed:
Use sideEffects instead of init

Co-authored-by: David Arenas <david.arenas@automattic.com>

luisherranz committed:
Move script registration to a general file

Co-authored-by: David Arenas <david.arenas@automattic.com>

luisherranz committed:
Move script registration to a general file

Co-authored-by: David Arenas <david.arenas@automattic.com>

luisherranz committed:
Move script registration to a general file

Co-authored-by: David Arenas <david.arenas@automattic.com>

luisherranz committed:
Add `defer` to the interactivity scripts

Co-authored-by: David Arenas <david.arenas@automattic.com>

luisherranz committed:
Add `defer` to the interactivity scripts

Co-authored-by: David Arenas <david.arenas@automattic.com>

luisherranz committed:
Add `defer` to the interactivity scripts

Co-authored-by: David Arenas <david.arenas@automattic.com>

luisherranz committed:
Revert changes of the image block

Co-authored-by: David Arenas <david.arenas@automattic.com>

luisherranz committed:
Revert changes of the image block

Co-authored-by: David Arenas <david.arenas@automattic.com>

luisherranz committed:
Revert changes of the image block

Co-authored-by: David Arenas <david.arenas@automattic.com>

luisherranz committed:
Fix init import name

Co-authored-by: David Arenas <david.arenas@automattic.com>

luisherranz committed:
Fix init import name

Co-authored-by: David Arenas <david.arenas@automattic.com>

luisherranz committed:
Fix init import name

Co-authored-by: David Arenas <david.arenas@automattic.com>

SantosGuillamot committed:
Add experimental setting to use Interactivity API

SantosGuillamot committed:
Add basic `interactivity.js` file

SantosGuillamot committed:
Enqueue old script/interactivity API conditionally

SantosGuillamot committed:
Add `tick()` function until Preact bug is solved

SantosGuillamot committed:
Add code to handle the overlayMenu

SantosGuillamot committed:
Add Interactivity API to submenu block

SantosGuillamot committed:
Add Interactivity API to page list block

gziolo commented on lib/experiments-page.php:
```diff
@@ -101,6 +101,18 @@ function gutenberg_initialize_experiments_settings() {
 		)
 	);
 
+	add_settings_field(
```
Should this experiment also cover other blocks that use Interactivity API?

SantosGuillamot replied:
To be honest, I am not sure if it should be all together or each block independently. For example, I can imagine the navigation block much easier to test than the [Block Behaviors](https://github.com/WordPress/gutenberg/issues/50029). I'm inclined to keep the discussions separate, but happy to change it if you consider it.

gziolo replied:
We can change it later if necessary 👍🏻

gziolo commented on packages/block-library/src/navigation/index.php:
```diff
@@ -561,13 +567,30 @@ function render_block_core_navigation( $attributes, $content, $block ) {
+		if ( $is_interactivity_api_enabled ) {
+			wp_enqueue_script(
+				'interactivity-navigation',
+				plugins_url( '../interactive-blocks/navigation.min.js', __FILE__ ),
```
There is also `__DIR__` constant that might work here better:
```suggestion
				plugins_url( './interactive-blocks/navigation.min.js', __DIR__ ),
```

SantosGuillamot replied:
I just changed it 🙂

gziolo commented on packages/block-library/src/navigation/index.php:
```diff
@@ -561,13 +567,30 @@ function render_block_core_navigation( $attributes, $content, $block ) {
 	$has_old_responsive_attribute = ! empty( $attributes['isResponsive'] ) && $attributes['isResponsive'];
 	$is_responsive_menu           = isset( $attributes['overlayMenu'] ) && 'never' !== $attributes['overlayMenu'] || $has_old_responsive_attribute;
 	$should_load_view_script      = ! wp_script_is( 'wp-block-navigation-view' ) && ( $is_responsive_menu || $attributes['openSubmenusOnClick'] || $attributes['showSubmenuIcon'] );
```
`wp_script_is` checks whether the script got previously enqueued. In effect it would be a different check when the Interactivity API is enabled.

SantosGuillamot replied:
Changed it as well 🙂 Thank you!

gziolo reviewed (commented):
I took it for a spin, and the version built on top of Interactivity API seems to work the same as the old one. That's impressive, given that the focus is managed manually 😄 

I was a bit surprised that the existing implementation of the overlay doesn't handle clicking outside of the menu as closing the dialog. Another missing feature was the using arrow up/down to move the focus between items in the submenu. Anyway, taking all that into account, it seems like both implementations have the same set of features.

See the recorded screencast with the existing behavior in the block editor when using menus and dialogs:

https://user-images.githubusercontent.com/699132/234262100-6a12f871-b829-4358-aa79-c7e4f62eb854.mov

I'm not saying we need to match that on the front end, but it is something I found surprising.

gziolo commented on packages/block-library/src/navigation/index.php:
```diff
@@ -793,35 +817,103 @@ function render_block_core_navigation( $attributes, $content, $block ) {
-				<div class="wp-block-navigation__responsive-close" tabindex="-1" data-micromodal-close>
-					<div class="wp-block-navigation__responsive-dialog" aria-label="%8$s">
-							<button %4$s data-micromodal-close class="wp-block-navigation__responsive-container-close">%10$s</button>
+	if ( $is_interactivity_api_enabled ) {
```
Would it be possible to use the tag processor instead to inject directives necessary for Interactivity API? This way we would need only a single markup definition for both, and the rest would be conditionally applied when the Interactivity API is enabled. I might be missing something that validates having two copies though.

SantosGuillamot replied:
I can definitely try that. It's something I thought but I wasn't sure what is better:
* Having the markup with the directives as it will be once/if it is not experimental anymore. I feel it is easier to understand it.
* Using the tag processor to avoid code duplication.
As I say, I don't have a strong opinion and I am not sure which one is better. I'd love to hear your thoughts.

gziolo replied:
Let's ignore my comment for now. It isn't simple to compare what changes between the new and the old version, but it's also a temporary approach for the testing period where we plan to validate the new approach. In a few weeks, we will most likely remove the old version and this way the process will be straightforward.

SantosGuillamot replied:
I've just pushed some commits to implement it using the tag processor, as you suggested, to be able to compare. It certainly looks cleaner, and we don't have to modify so many blocks.

**With the tag processor**

* [Navigation block](https://github.com/WordPress/gutenberg/blob/1d0472db69a1d2fae021f6591a5eb7163318dab2/packages/block-library/src/navigation/index.php#L564-L592): We just have to add the scripts conditionally (maybe this can be moved as well).
* [Navigation submenu block](https://github.com/WordPress/gutenberg/blob/1d0472db69a1d2fae021f6591a5eb7163318dab2/packages/block-library/src/navigation-submenu/index.php): It doesn't change at all.
* [Page list block](https://github.com/WordPress/gutenberg/blob/1d0472db69a1d2fae021f6591a5eb7163318dab2/packages/block-library/src/page-list/index.php): It doesn't change at all.

The benefits I see are that it looks much simpler and it might be more compatible with other changes made to the navigation block.

**Without the tag processor**

* [Navigation block](https://github.com/WordPress/gutenberg/blob/e63c002f0a8a596ee587f94f6d905d218d0785ed/packages/block-library/src/navigation/index.php#L822-L927): Apart from adding the scripts conditionally, we have to include more conditionals to change the markup, and duplicate it in most parts.
* [Navigation submenu block](https://github.com/WordPress/gutenberg/blob/e63c002f0a8a596ee587f94f6d905d218d0785ed/packages/block-library/src/navigation-submenu/index.php#L127-L301): We have to include conditionals to change the markup, and duplicate it.
* [Page list block](https://github.com/WordPress/gutenberg/blob/e63c002f0a8a596ee587f94f6d905d218d0785ed/packages/block-library/src/page-list/index.php#L203L268): We have to include conditionals to change the markup, and duplicate it.

The benefits I see are that it is clearer how to use directives and the old markup should be removed once/if it is not experimental anymore.

Let me know what you think and which option you prefer 🙂

gziolo replied:
I responded in the comment with general feedback in favor of using the tag processor: https://github.com/WordPress/gutenberg/pull/50041#issuecomment-1527427568.

SantosGuillamot committed:
Change code to be more declarative and simple

SantosGuillamot committed:
Use `__DIR__` instead of `__FILE__`

SantosGuillamot committed:
Change loading script conditional

SantosGuillamot committed:
Fix issues momentarily

SantosGuillamot committed:
Fix some issues with focus

SantosGuillamot committed:
Add tabindex to modal

SantosGuillamot committed:
Revert to current navigation block

SantosGuillamot committed:
Add directives through a filter

SantosGuillamot committed:
Dont add latest modifications to navigation block

SantosGuillamot committed:
Ensure focus doesn't get inside menu after closing

gziolo commented on packages/block-library/src/navigation/interactivity.js:
```diff
@@ -0,0 +1,121 @@
+				initModal: async ( { context, ref } ) => {
+					if ( context.isMenuOpen ) {
+						const focusableElements = ref.querySelectorAll(
+							'a[href], button:not([disabled]), textarea, input[type="text"], input[type="radio"], input[type="checkbox"], select'
```
The same list in Micromodal:
https://github.com/ghosh/Micromodal/blob/2a316150d12b3f3a0540762ee77dcb7247c6518a/lib/src/index.js#LL4C3-L16C4

SantosGuillamot committed:
Add conditionals and change function names

SantosGuillamot committed:
Use `gutenberg_is_experiment_enabled` function

gziolo commented:
I wanted to highlight the comment https://github.com/WordPress/gutenberg/pull/50041#discussion_r1179373768 from @SantosGuillamot so it's visible for everyone. In particular this part:

> **With the tag processor**
> 
> - [Navigation block](https://github.com/WordPress/gutenberg/blob/1d0472db69a1d2fae021f6591a5eb7163318dab2/packages/block-library/src/navigation/index.php#L564-L592): We just have to add the scripts conditionally (maybe this can be moved as well).
> - [Navigation submenu block](https://github.com/WordPress/gutenberg/blob/1d0472db69a1d2fae021f6591a5eb7163318dab2/packages/block-library/src/navigation-submenu/index.php): It doesn't change at all.
> - [Page list block](https://github.com/WordPress/gutenberg/blob/1d0472db69a1d2fae021f6591a5eb7163318dab2/packages/block-library/src/page-list/index.php): It doesn't change at all.
>
> The benefits I see are that it looks much simpler and it might be more compatible with other changes made to the navigation block.

Thank you so much for exploring the alternative approach based on the tag processor. As noted in my previous comment, it didn't seem as important as we are only trying an experiment that will run for a couple of weeks. Obviously, we hope to adopt the new way based on Interactivity API, but we need to ensure everything is polished, we get enough feedback from folks and that's why we will keep the alternative solution behind the experimental feature flag.

Now, that I see the implementation, I realized that we can handle the experiment exclusively with WP filters. The code uses already `render_block_core/navigation` hook, but we could also use [`block_type_metadata`](https://developer.wordpress.org/reference/hooks/block_type_metadata/) to replace the JavaScript files. This way, there will be no visible change in the original implementation of the Navigation block. I like it firstly because it doesn't obscure the current implementation, and we ensure it won't leak to WordPress Core during the merge process. Another very important reason I find the current proposal so appealing is that it presents how third-party projects will be able to enhance core blocks with additional functionality.

gziolo commented:
Possible follow-up task:

I was searching for existing e2e tests that cover the Navigation block, but I could locate anything that would verify the behavior on the front end when there is the menu icon that opens the full page overlay. The closest test that I located which could be the starting point is:

https://github.com/WordPress/gutenberg/blob/de44e4413fb40e5dcc79faeb3cd26380f9b4446d/test/e2e/specs/editor/blocks/navigation.spec.js#L55-L94

The nice part of having an e2e test that verifies the interactive behavior is that we could run it again the old and the new version on CI until we switch to the final approach.

gziolo committed:
Move and refactor the interactive scritps registration

gziolo committed:
Move and refactor the interactive scritps registration

gziolo committed:
Fix code style violations

gziolo committed:
Fix code style violations

gziolo committed:
Use `wp-interactivity-` prefix for script handles

gziolo committed:
Use `wp-interactivity-` prefix for script handles

gziolo committed:
Improve the matcher for side effects in `package.json`

gziolo committed:
Improve the matcher for side effects in `package.json`

SantosGuillamot committed:
Merge branch 'add-interactivity-runtime' into navigation-block-with-interactivity-api

SantosGuillamot committed:
Enqueue interactivity scripts through hook

SantosGuillamot committed:
Revert `client-assets.php` file

SantosGuillamot committed:
Adapt to latest changes in base branch

DAreRodz committed:
Add custom useSignalEffect

DAreRodz committed:
Add custom useSignalEffect

SantosGuillamot committed:
Merge branch 'add-interactivity-runtime' into navigation-block-with-interactivity-api

SantosGuillamot committed:
Move role attribute to selectors

gziolo commented:
Here is the list of issues with PHP reported on CI:

<img width="775" alt="Screenshot 2023-05-04 at 15 55 47" src="https://user-images.githubusercontent.com/699132/236229287-8d47465d-e245-4a95-bd39-255df109633b.png">

DAreRodz committed:
Call `init` after `store` has been initialized

DAreRodz committed:
Call `init` after `store` has been initialized

SantosGuillamot committed:
Merge branch 'add-interactivity-runtime' into navigation-block-with-interactivity-api

SantosGuillamot committed:
Change focusable elements

SantosGuillamot committed:
Add namespaces to context

SantosGuillamot committed:
Move PHP file to interactivity-api folder

SantosGuillamot committed:
Add comments with the markup to the PHP filter

SantosGuillamot committed:
Remove micromodal attributes

gziolo commented on lib/experimental/interactivity-api/navigation-block-interactivity.php:
```diff
@@ -0,0 +1,214 @@
+ */
+
+$gutenberg_experiments = get_option( 'gutenberg-experiments' );
+if ( gutenberg_is_experiment_enabled( 'gutenberg-interactivity-api-navigation-block' ) ) {
```
It's a matter of preference, but you could also put instead the same check before including the file in `load.php`. This way you could avoid additional tab in nearly every line in this file.

SantosGuillamot replied:
That makes sense 🙂 I'll change it

gziolo commented on tools/webpack/blocks.js:
```diff
@@ -221,6 +221,8 @@ module.exports = [
 		entry: {
+			navigation:
+				'./packages/block-library/src/navigation/interactivity.js',
 			// blockname: './packages/block-library/src/blockname/interactivity.js',
```
It's fine to remove this comment now:
```suggestion
```

gziolo committed:
Update lib/experimental/interactivity-api/script-loader.php

Co-authored-by: Weston Ruter <westonruter@google.com>

gziolo committed:
Plugin: Ensure that translations are set correctly when overriding scripts
This replicates the same logic in WordPress core:
https://github.com/WordPress/wordpress-develop/blob/a5f3087f6b5d9c52dbe67ed247165dc32427c57d/src/wp-includes/script-loader.php#L306-L308

SantosGuillamot committed:
Remove unnecessary comment in webpack config

Co-authored-by: Greg Ziółkowski <grzegorz@gziolo.pl>

SantosGuillamot committed:
Merge branch 'add-interactivity-runtime' into navigation-block-with-interactivity-api

SantosGuillamot committed:
Clean up code to fix PHP coding standards issues

SantosGuillamot committed:
Remove extra space

SantosGuillamot committed:
Merge branch 'trunk' into navigation-block-with-interactivity-api

scruffian commented:
Thanks for working on this. I am curious about a couple of things:

1. What does does this replace? I assume it's the contents of `view.js` and `view-modal.js`? Is that right?
2. What advantages does this approach offer over how we currently do it?

Thanks again!

SantosGuillamot commented:
First of all, sorry for the lack of a proper description. I wanted to cover those questions and more there. I didn't find the time to work on it yet. I'll try to do it today.

> What does this replace? I assume it's the contents of view.js and view-modal.js? Is that right?

That's it 🙂 The idea of this pull request is to use [the Interactivity API](https://make.wordpress.org/core/2023/03/30/proposal-the-interactivity-api-a-better-developer-experience-in-building-interactive-blocks/) (as an experiment) instead of `micromodal` library for the navigation block.

> What advantages does this approach offer over how we currently do it?

In terms of functionalities, it just aims to replicate the exact same behaviors but using [the Interactivity API](https://make.wordpress.org/core/2023/03/30/proposal-the-interactivity-api-a-better-developer-experience-in-building-interactive-blocks/).

This pull request would come as part of [its proposal](https://make.wordpress.org/core/2023/03/30/proposal-the-interactivity-api-a-better-developer-experience-in-building-interactive-blocks/). The idea is to start using the Interactivity API in more parts of Gutenberg. Some examples could be [the lightbox in Image block](https://github.com/WordPress/gutenberg/pull/50373), the file block, or an accordion block among others. With this in mind, this would be the main benefits:
* If many blocks use the Interactivity API, they would share the same JS files so it would translate into less KBs sent to the client if you are using more than one block.
* It wouldn't depend on external libraries like `micromodal`.
* It would help to test the Interactivity API proposal and serve as an example.

scruffian commented:
Thanks for the details, very interesting. It would be great to see how many kB this PR would save, but I'm sure you are already planning to add that to the description :)

SantosGuillamot committed:
Add tests to the navigation block interactivity

SantosGuillamot committed:
Modify `saveSiteEditorEntities` function to wait

SantosGuillamot committed:
Fix page-list tests

gziolo commented:
E2E tests are now in the `trunk` after #50462 landed. They were extracted from this PR and iterated upon. I will test this PR to see if it is ready to land for the Gutenberg plugin RC1 planned for tomorrow. Everything is hidden behind the feature flag so it should be a safe rollout that will allow us a broader testing for Interactivity API on real sites.

gziolo committed:
Remove e2e tests that are already included in trunk

gziolo commented:
I tested this PR, and everything works as advertised.

I can enable/disable the experiment with the setting:

<img width="846" alt="Screenshot 2023-05-09 at 14 51 05" src="https://github.com/WordPress/gutenberg/assets/699132/031f8dc9-af5c-4367-b33c-964b70914405">

I see only 3 scripts loaded for the Navigation block (view script, interactivity runtime, and vendors). There is one issue to resolve for the view script, as `defer` is not set. Somehow, the scripts are executed in the correct order, but it won't be guaranteed. Maybe webpack will make it work, but I assume we want to have all interactivity scripts working with `defer`.

<img width="1557" alt="Screenshot 2023-05-09 at 14 48 40" src="https://github.com/WordPress/gutenberg/assets/699132/46c7a68c-ec3c-4bc7-9c4c-d5cef0a6088e">

The menu icon looks like before (without the experiment enabled):

<img width="842" alt="Screenshot 2023-05-09 at 14 50 12" src="https://github.com/WordPress/gutenberg/assets/699132/b29d226c-8eac-4551-af9d-01c6fe3c5085">

The menu overlay behaves as before (without the experiment enabled):

<img width="401" alt="Screenshot 2023-05-09 at 14 50 20" src="https://github.com/WordPress/gutenberg/assets/699132/684bc075-afd8-4300-87a5-42fec2f0997c">

gziolo commented:
I discovered an issue with the case where you didn't opt for the modal overlay to show the menu:

<img width="277" alt="Screenshot 2023-05-09 at 16 58 04" src="https://github.com/WordPress/gutenberg/assets/699132/4e17ee54-9c6e-4371-b7fc-0d4c45a63a5a">

<img width="515" alt="Screenshot 2023-05-09 at 16 58 29" src="https://github.com/WordPress/gutenberg/assets/699132/1329a834-0f29-47a2-9ca7-6ec62bcb37bb">

In that case, the code that modifies the HTML of the Navigation block won't ever find the button that opens the menu, and will finish processing HTML without any further updates. We could fix it with something close to:

```diff
diff --git a/lib/experimental/interactivity-api/navigation-block-interactivity.php b/lib/experimental/interactivity-api/navigation-block-interactivity.php
index 0453c8ed1a..3d363db4df 100644
--- a/lib/experimental/interactivity-api/navigation-block-interactivity.php
+++ b/lib/experimental/interactivity-api/navigation-block-interactivity.php
@@ -58,6 +58,7 @@ function gutenberg_block_core_navigation_add_directives_to_markup( $block_conten
 	if ( $w->next_tag( 'nav' ) ) {
 		$w->set_attribute( 'data-wp-island', '' );
 		$w->set_attribute( 'data-wp-context', '{ "core": { "navigation": { "isMenuOpen": false, "overlay": true, "roleAttribute": "" } } }' );
+		$w->set_bookmark( 'nav-element' );
 	};
 
 	// Add directives to the open menu button.
@@ -70,6 +71,28 @@ function gutenberg_block_core_navigation_add_directives_to_markup( $block_conten
 		$w->set_attribute( 'data-wp-on.click', 'actions.core.navigation.openMenu' );
 		$w->set_attribute( 'data-wp-on.keydown', 'actions.core.navigation.handleMenuKeydown' );
 		$w->remove_attribute( 'data-micromodal-trigger' );
+	} else {
+		// If the open modal button not found, we handle submenus immediately.
+		$w->seek( 'nav-element' );
+
+		// Add directives to the menu container.
+		if ( $w->next_tag(
+			array(
+				'tag_name'   => 'UL',
+				'class_name' => 'wp-block-navigation__container',
+			)
+		) ) {
+			$w->set_attribute( 'data-wp-class.is-menu-open', 'context.core.navigation.isMenuOpen' );
+			$w->set_attribute( 'data-wp-bind.aria-hidden', '!context.core.navigation.isMenuOpen' );
+			$w->set_attribute( 'data-wp-effect', 'effects.core.navigation.initModal' );
+			$w->set_attribute( 'data-wp-on.keydown', 'actions.core.navigation.handleMenuKeydown' );
+			$w->set_attribute( 'data-wp-on.focusout', 'actions.core.navigation.handleMenuFocusout' );
+			$w->set_attribute( 'tabindex', '-1' );
+		};
+
+		gutenberg_block_core_navigation_add_directives_to_submenu( $w );
+
+		return $w->get_updated_html();
 	}
 
 	// Add directives to the menu container.
@@ -124,7 +147,7 @@ function gutenberg_block_core_navigation_add_directives_to_markup( $block_conten
 	// Submenus.
 	gutenberg_block_core_navigation_add_directives_to_submenu( $w );
 
-	return (string) $w;
+	return $w->get_updated_html();
 };
 
 /**
@@ -209,7 +232,7 @@ add_filter(
 				gutenberg_url( 'build/block-library/interactive-blocks/navigation.min.js' ),
 				array( 'wp-interactivity-runtime' )
 			);
-			$metadata['viewScript'] = "array('wp-block-navigation-view')";
+			$metadata['viewScript'] = array( 'wp-block-navigation-view' );
 		}
 		return $metadata;
 	},
```

However, it doesn't work correctly, as the view script code most likely assumes there is always an overlay. The printed HTML with the diff applied:

<img width="1595" alt="Screenshot 2023-05-10 at 10 04 33" src="https://github.com/WordPress/gutenberg/assets/699132/db13df24-94f2-49dc-b1a9-e1e19c83c445">

gziolo committed:
Fix the navigation menu when the overlay is not present

gziolo committed:
Ensure that all scripts using Intereactivity API have defer attribute

gziolo committed:
Remove unused param in the callback used by hook

gziolo commented:
I think I resolved both issues reported earlier:
- https://github.com/WordPress/gutenberg/pull/50041/commits/c6ee68acc08ace0db9cd237e12b6fefec84faf22 fixes the Navigation block without overlay described in https://github.com/WordPress/gutenberg/pull/50041#issuecomment-1541537641.
- https://github.com/WordPress/gutenberg/pull/50041/commits/15d354635375f2ca076c779d17e7282c423e1c95 fixes the issue I raised in https://github.com/WordPress/gutenberg/pull/50041#issuecomment-1540070946 about using `defer` with all scripts that consume Interactivity API runtime.

gziolo reviewed (approved):
I think this PR is now in good shape. I would appreciate some help with testing my recent changes.

gziolo commented:
Test failures from CI jobs for checking PHP unit tests fail now because of the regression introduced in `trunk` with ccd529d7a6af0e233c2bef03dde214ca48165c75.

DAreRodz commented:
> I would appreciate some help with testing my recent changes.

Taking a look at them, @gziolo.

cbravobernal reviewed (approved):
LGTM.
Tested and worked as expected. Keyboard navigation works smoothly.
Congrats!

DAreRodz commented:
@gziolo, I've tested the changes you mentioned in https://github.com/WordPress/gutenberg/pull/50041#issuecomment-1541647248, and the issues seem to be solved now. 👍

gziolo committed:
Merge remote-tracking branch 'origin/trunk' into navigation-block-with-interactivity-api

gziolo commented on lib/experimental/interactivity-api/navigation-block-interactivity.php:
```diff
@@ -0,0 +1,240 @@
+		) ) {
+			$w->set_attribute( 'data-wp-class.is-menu-open', 'context.core.navigation.isMenuOpen' );
+			$w->set_attribute( 'data-wp-bind.aria-hidden', '!context.core.navigation.isMenuOpen' );
+			$w->set_attribute( 'data-wp-effect', 'effects.core.navigation.initModal' );
```
@SantosGuillamot, would it make sense to rename the effect to reflect the fact it works with the menu when there is no modal, too?

SantosGuillamot replied:
I renamed it to `initMenu` in [this commit](https://github.com/WordPress/gutenberg/pull/50670/commits/0ce72a818dc3252be365566e00c2b9f8c0556fb0).

gyurmey2 commented:
Version 15.9.20230511 returns 404 for the following files:
```
vendors.min.js
interactivity.min.js
navigation.min.js
```

gziolo commented:
It's very likely that those files aren't whitelisted for inclusion in the zip file attached to the release:
https://github.com/WordPress/gutenberg/blob/37d98396f63dd7463287fd9d2124a179a1f68458/bin/build-plugin-zip.sh#L81-L89

gziolo commented:
I opened https://github.com/WordPress/gutenberg/pull/50598 to fix the issue with missing scripts.
